### PR TITLE
Task

### PR DESCRIPTION
**Pull request by zwaldowski**
#### What's in this pull request?

See #64.

Introduces a `TaskType`, defining a cancellable `Future` over a `Result`.

`Task` (mirroring `Future`) is the main entry point into this API.

#### Testing

Cleans up some shared fixtures between Deferred and Task unit tests.

- [x] Improve test coverage of `Task` and its various and sundry extensions.
- [x] Improve test coverage of `FutureType where Value: ResultType` extensions.

The new additions have reasonably high test coverage.

#### API Changes

Changes are entirely additive to Deferred 2.0.

### CODE REVIEW

File: Sources/Result/ResultType.swift [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+//  Deferred
+//
+//  Created by Zachary Waldowski on 12/9/15.
+//  Copyright © 2014-2016 Big Nerd Ranch. Licensed under MIT.
+//
+
+/// A type that can exclusively represent either some result value of a
+/// successful computation or a failure with an error.
+public protocol ResultType: CustomStringConvertible, CustomDebugStringConvertible {
+    associatedtype Value
```
Comment (zwaldowski) [inline Sources/Result/ResultType.swift]: I invite improvements to the name of this associated type. Bike shedders, activate!

File: Sources/Task/ExistentialTask.swift [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        cancellation()
+    }
+}
+
+extension Task {
+    /// Create a task whose `upon(_:body:)` method uses the result of `base`.
+    public init<Task: FutureType where Task.Value: ResultType, Task.Value.Value == SuccessValue>(_ base: Task, cancellation: Cancellation) {
+        self.init(base.map {
+            Value(with: $0.extract)
+        }, cancellation: cancellation)
```
Comment (zwaldowski) [inline Sources/Task/ExistentialTask.swift]: This could/should be a `lazy` map if/when we resurrect #75.

File: Sources/Task/ExistentialTask.swift [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    /// Create a task whose `upon(_:body:)` method uses the result of `base`.
+    public init<Task: TaskType where Task.Value.Value == SuccessValue>(_ base: Task) {
+        self.init(base.map {
+            Value(with: $0.extract)
+        }, cancellation: base.cancel)
```
Comment (zwaldowski) [inline Sources/Task/ExistentialTask.swift]: Same above as with `lazy`.

File: Sources/Task/ExistentialTask.swift (hunk 2 above)
Comment (zwaldowski) [inline Sources/Task/ExistentialTask.swift]: Strike that, these should use the extension we define on `ResultType` in `ResultFuture.swift`.

File: Sources/Task/ResultFuture.swift [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        upon(Self.genericQueue, body: commonUponFailure(body))
+    }
+}
+
+extension Future where Value: ResultType {
+    /// Create a future having the same underlying task as `other`.
+    public init<Other: FutureType where Other.Value: ResultType, Other.Value.Value == Value.Value>(_ other: Other) {
+        self = other.map {
+            Value(with: $0.extract)
+        }
```
Comment (zwaldowski) [inline Sources/Task/ResultFuture.swift]: This could/should be a `lazy` map if/when we resurrect #75.

File: Sources/Task/ResultPromise.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import Result
+#endif
+
+extension PromiseType where Value: ResultType {
+    /// Completes the task with a successful `value`, or a thrown error.
+    ///
+    /// - seealso: `fill(_:)`
+    public func succeed(@autoclosure value: () throws -> Value.Value) -> Bool {
+        return fill(Value(value: value))
+    }
```
Comment (zwaldowski) [inline Sources/Task/ResultPromise.swift]: Like the `ResultType` initializer, I don't think this works like we want it to. Need to test if `future.succeed(try foo())` (or `try future.succeed(foo())`?) still requires the surrounding context be `throws`.

Comment (zwaldowski) [inline Sources/Task/ResultPromise.swift]: If it doesn't work, that's probably a compiler bug.

**Comment by davidahouse** [comment]
I wrote some introductory tests for Task today and so far this work looks good. One question is where you plan to go with the `TaskAccumulator` and `TaskCollections`? These feel like the same kind of thing in general with very different implementations (and of course the difference between if it cares about the result or not). Overall I like the thought of using `TaskCollections` as a way to get a single Task that embodies some combination of an array of tasks (all finished, all success, first finished, etc). But you might have a more grand vision planed. Please share. :)

**Comment by zwaldowski** [comment]
There's not yet an over-arching vision for it, just that they were both in our original distribution of `Task`. We have definitely considered/weighed the different paths you'd want. The distinctions are fairly subtle.

Currently, there is:
* With `[Task<T>]`, use `joinedValues` to know when everything _finishes_, getting a `Future<[Result<T>]>`.
   - We could perhaps shadow this to make the result a cancellable `Task<[T]>`.
* With `[Task<T>]`, use `joinedTasks` to know when everything _succeeds_, getting a `Task<Void>`.
* With an incremental list of any `Task` (for example, a Core Data store that vends `Task`s over time), use `TaskAccumulator` to know when everything currently in-flight has _finished_.

@jgallagher can speak better to the idea of `TaskAccumulator`, and weigh in on whether we need it.

**Comment by davidahouse** [comment]
Cool sounds good! I'll try that joinedValues. That is one thing I was looking for and didn't find.

File: Sources/Task/ResultPromise.swift (hunk 5 above)
Comment (zwaldowski) [inline Sources/Task/ResultPromise.swift]: Seems fine to me. You can't `throw` inside the parens, though, which is to be expected.

File: Sources/Result/ResultType.swift (hunk 1 above)
Comment (zwaldowski) [inline Sources/Result/ResultType.swift]: On hold until Swift 3

**Comment by zwaldowski** [comment]
I'm splitting `NSURLSession` bits into #109 to maybe get this train moving; it was the piece lacking testing entirely.

File: Sources/Task/ResultPromise.swift (hunk 5 above)
Comment (zwaldowski) [inline Sources/Task/ResultPromise.swift]: I've identified other places `@autoclosure` breaks for this, like the `ResultType` initializer. Swift 3 it is…

**Comment by zwaldowski** [comment]
Coverage is up to 75%. Covering `TaskCollections.swift` and `IgnoringTask.swift` are my last big-ticket items, if anyone wanted to hit those before I can.

Tasking off WIP; ready for review.

File: Configurations/Base.xcconfig [hunk 6]
```diff
@@ -39,8 +41,6 @@ TVOS_DEPLOYMENT_TARGET = 9.0
 ENABLE_BITCODE = YES
 ENABLE_BITCODE[sdk=macosx*] = NO
 
-CODE_SIGN_IDENTITY[sdk=macosx*] = -
-CODE_SIGN_IDENTITY[sdk=iphoneos*] = iPhone Developer
-CODE_SIGN_IDENTITY[sdk=appletvos*] = iPhone Developer
+CODE_SIGN_IDENTITY =
```
Comment (zwaldowski) [inline Configurations/Base.xcconfig]: Drop this, only for Xcode 8/Swift 3

**Pull request merged**